### PR TITLE
[Enhancement] TensorRT Anchor generator plugin

### PR DESCRIPTION
**Pull request by grimoire**
Screenshot from 2022-06-25 14-36-12
Screenshot from 2022-06-25 14-36-07

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                         Memcpy DtoD (Device -> Device)         0.00%       0.000us         0.00%       0.000us       0.000us      73.000us        47.10%      73.000us      73.000us             1  
void cuFillLayer::fill<int>(cuFillLayer::KernelArgs<...         0.00%       0.000us         0.00%       0.000us       0.000us      38.000us        24.52%      38.000us      19.000us             2  
void cuEltwise::eltwise<cuEltwise::SimpleAlgo<float,...         0.00%       0.000us         0.00%       0.000us       0.000us      24.000us        15.48%      24.000us      24.000us             1  
void genericReformat::copyPackedKernel<float, float,...         0.00%       0.000us         0.00%       0.000us       0.000us      10.000us         6.45%      10.000us       3.333us             3  
void cuSliceLayer::naiveSlice<float, (cuSliceLayer::...         0.00%       0.000us         0.00%       0.000us       0.000us       5.000us         3.23%       5.000us       2.500us             2  
                               generatedNativePointwise         0.00%       0.000us         0.00%       0.000us       0.000us       3.000us         1.94%       3.000us       1.500us             2  
void cuSliceLayer::naiveSlice<float, (cuSliceLayer::...         0.00%       0.000us         0.00%       0.000us       0.000us       2.000us         1.29%       2.000us       1.000us             2  
                                        cudaEventRecord         5.46%     440.000us         5.46%     440.000us     440.000us       0.000us         0.00%       0.000us       0.000us             1  
                                       cudaLaunchKernel        36.58%       2.948ms        36.58%       2.948ms     245.667us       0.000us         0.00%       0.000us       0.000us            12  
void cuCastLayer::cast<int, float, float, false, fal...         0.00%       0.000us         0.00%       0.000us       0.000us       0.000us         0.00%       0.000us       0.000us             2  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CUDA time total: 155.000us
```

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
void trt_grid_priors_kernel<float>(float const*, flo...         0.00%       0.000us         0.00%       0.000us       0.000us      11.000us        84.62%      11.000us      11.000us             1  
void cuSliceLayer::naiveSlice<int, (cuSliceLayer::Mo...         0.00%       0.000us         0.00%       0.000us       0.000us       2.000us        15.38%       2.000us       1.000us             2  
                                        cudaEventRecord         0.00%       5.000us         0.00%       5.000us       5.000us       0.000us         0.00%       0.000us       0.000us             1  
                                       cudaLaunchKernel       100.00%        2.366s       100.00%        2.366s     788.507ms       0.000us         0.00%       0.000us       0.000us             3  
                                  cudaDeviceSynchronize         0.00%      36.000us         0.00%      36.000us      36.000us       0.000us         0.00%       0.000us       0.000us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CUDA time total: 13.000us

```

### CODE REVIEW

File: mmdeploy/codebase/mmdet/core/anchor.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+grid_priors_trt = GridPriorsTRTOp.apply
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    func_name='mmdet.core.anchor.anchor_generator.'
+    'AnchorGenerator.single_level_grid_priors',
+    backend='tensorrt')
+def anchorgenerator__single_level_grid_priors__trt(ctx,
```
Comment (AllentDan) [inline mmdeploy/codebase/mmdet/core/anchor.py:68]: Need docstring

File: mmdeploy/codebase/mmdet/core/anchor.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            feat_h,
+            value_t=torch.tensor([0], dtype=torch.long),
+        )
+        zero_w = g.op(
+            'ConstantOfShape',
+            feat_w,
+            value_t=torch.tensor([0], dtype=torch.long),
+        )
+        return g.op(
+            'mmdeploy::GridPriorsTRT',
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/core/anchor.py:53]: could update the tensorrt ops doc: https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/ops/tensorrt.md#tensorrt-ops

File: mmdeploy/codebase/mmdet/core/anchor.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+def anchorgenerator__single_level_grid_priors__trt(ctx,
+                                                   self,
+                                                   featmap_size,
+                                                   level_idx,
+                                                   dtype=torch.float32,
+                                                   device='cuda'):
+    # generate origin func for forward
+
+    feat_h, feat_w = featmap_size
+    if isinstance(feat_h, int) and isinstance(feat_w, int):
```
[the remark this answers is not in the document]
Comment (grimoire) [reply mmdeploy/codebase/mmdet/core/anchor.py:77]: Yes, if size is fixed, we can just output the pre-computed anchors.

File: docs/en/ops/tensorrt.md [hunk 4]
```diff
@@ -363,3 +369,39 @@ Batched rotated NMS with a fixed number of output bounding boxes.
...
+| `int` | `stride_w` | The stride of the feature width.  |
+| `int` | `stride_h` | The stride of the feature height. |
+
+#### Inputs
+
+<dl>
+<dt><tt>inputs[0]</tt>: T</dt>
+<dd>The base anchors; 2-D tensor with shape [num_base_anchor, 4].</dd>
+<dt><tt>inputs[1]</tt>: TAny</dt>
+<dd>height provider; 1-D tensor with shape [featmap_height]. The data will never been used.</dd>
```
Comment (RunningLeon) [inline docs/en/ops/tensorrt.md:392]: What does it mean by `The data will never be used`?

Comment (grimoire) [reply docs/en/ops/tensorrt.md:392]: The tensor is used to provide its shape, the data will not be read in the plugin.

**Comment by RunningLeon** [comment]
@grimoire may fix conflicts

**Review by AllentDan — APPROVED** [review]
LGTM

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**